File: tests/legacy/server/notifications.ts
import type { PushSubscription } from 'web-push'
import type { ChatQuestion } from '../../../shared/chats.ts'
import type { Config } from './config.ts'
import type { Store } from './store.ts'
import { Buffer } from 'node:buffer'
import { createHash } from 'node:crypto'
import webpush from 'web-push'
import { z } from 'zod'
import { AppError } from './errors.ts'
import { McpVault } from './mcp-vault.ts'

const subscriptionInput = z.object({
  endpoint: z.string().url().max(4096),
  keys: z.object({ p256dh: z.string().regex(/^[\w-]+={0,2}$/).max(100), auth: z.string().regex(/^[\w-]+={0,2}$/).max(30) }),
})
interface Delivery { subscriptionId: string, chatId: string, questionId: string, attempts: number, nextAt: number, expiresAt: number }
export class Notifications {
  private vault: McpVault
  private timer?: ReturnType<typeof setInterval>
  private pending?: Promise<void>
  constructor(private store: Store, private config: Config, private send = webpush.sendNotification) {
    this.vault = new McpVault(store, config.dataDir)
  }

  private keys() {
    let keys = this.vault.get<{ publicKey: string, privateKey: string }>('push-vapid')
    if (!keys) {
      keys = webpush.generateVAPIDKeys()
      this.vault.set('push-vapid', keys)
    }
    return keys
  }

  configuration() { return { publicKey: this.keys().publicKey } }
  registered(id: string) { return !!this.store.kv(`push-device:${id}`) }
  subscribe(input: unknown) {
    const subscription = subscriptionInput.parse(input)
    const url = new URL(subscription.endpoint)
    // Only browser push services may receive server-side requests. In particular,
    // never turn this authenticated endpoint into a proxy to internal services.
    const allowed = ['fcm.googleapis.com', 'updates.push.services.mozilla.com', 'web.push.apple.com']
    const suffixes = ['.push.services.mozilla.com', '.notify.windows.com']
    if (url.protocol !== 'https:' || url.username || url.password || url.port || url.hash || (!allowed.includes(url.hostname) && !suffixes.some(suffix => url.hostname.endsWith(suffix))))
      throw new AppError(400, 'This browser push service is not supported.')
    if (Buffer.from(subscription.keys.p256dh, 'base64url').length !== 65 || Buffer.from(subscription.keys.auth, 'base64url').length !== 16)
      throw new AppError(400, 'Invalid notification subscription keys.')
    const id = createHash('sha256').update(subscription.endpoint).digest('hex')
    if (!this.registered(id) && this.store.keys('push-device:').length >= 50)
      throw new AppError(409, 'Too many notification devices are registered.')
    this.store.transaction(() => {
      this.vault.set(`push-device:${id}`, subscription)
      this.store.set(`push-device:${id}`, { createdAt: Date.now() })
    })
    return { id }
  }

  unsubscribe(id: string) {
    this.vault.delete(`push-device:${id}`)
    this.store.delete(`push-device:${id}`)
    for (const { key, data } of this.store.keys('push-outbox:')) {
      if (data.subscriptionId === id)
        this.store.delete(key)
    }
    return { ok: true }
  }

  enqueue(question: ChatQuestion) {
    for (const { key } of this.store.keys('push-device:')) {
      const subscriptionId = key.slice('push-device:'.length)
      const id = `push-outbox:${question.id}:${subscriptionId}`
      if (!this.store.kv(id))
        this.store.set(id, { subscriptionId, questionId: question.id, chatId: question.chatId, attempts: 0, nextAt: Date.now(), expiresAt: Date.now() + 3600000 } satisfies Delivery)
    }
  }

  async flush() {
    if (this.pending)
      return this.pending
    this.pending = this.deliver().finally(() => {
      this.pending = undefined
    })
    return this.pending
  }

  private async deliver() {
    for (const { key, data } of this.store.keys('push-outbox:')) {
      const delivery = data as Delivery
      if (delivery.nextAt > Date.now())
        continue
      const question = this.store.kv<ChatQuestion>(`chat-question:${delivery.chatId}:${delivery.questionId}`)
      const subscription = this.vault.get<PushSubscription>(`push-device:${delivery.subscriptionId}`)
      if (!subscription || question?.status !== 'pending' || delivery.expiresAt < Date.now()) {
        this.store.delete(key)
        continue
      }
      try {
        const subject = this.config.publicUrl.startsWith('https:') ? this.config.publicUrl : 'mailto:notifications@example.com'
        await this.send(subscription, JSON.stringify({ title: 'Your agent has a question', body: 'Open the chat to answer.', chatId: delivery.chatId, questionId: delivery.questionId }), { TTL: 3600, urgency: 'high', timeout: 5000, vapidDetails: { subject, ...this.keys() } })
        this.store.delete(key)
      }
      catch (error) {
        const status = (error as { statusCode?: number }).statusCode
        if (status === 404 || status === 410) {
          this.unsubscribe(delivery.subscriptionId)
          continue
        }
        // Store no endpoint, response body, question text, or subscription secret in logs.
        this.store.set(key, { ...delivery, attempts: delivery.attempts + 1, nextAt: Date.now() + Math.min(300000, 10000 * 2 ** delivery.attempts) })
      }
    }
  }

  start() {
    this.timer ??= setInterval(() => {
      void this.flush().catch(() => {})
    }, 1000)
    this.timer.unref()
  }

  async close() {
    clearInterval(this.timer)
    await this.pending
  }
}

File: backend/src/notifications.rs
use crate::{
    auth::hex_digest,
    config::now,
    error::{Error, Result},
    service::Service,
    store::Db,
    validation::text,
};
use base64::{Engine, engine::general_purpose::URL_SAFE_NO_PAD};
use p256::elliptic_curve::sec1::ToEncodedPoint;
use serde_json::{Value, json};
use std::{sync::Arc, time::Duration};
use tokio::sync::Mutex;
use web_push::{
    ContentEncoding, HyperWebPushClient, SubscriptionInfo, Urgency, VapidSignatureBuilder,
    WebPushClient, WebPushError, WebPushMessageBuilder,
};
#[derive(Clone, Default)]
pub struct Notifications {
    delivery: Arc<Mutex<()>>,
}
pub fn enqueue(db: &Db<'_>, question: &Value) -> Result<()> {
    for (key, _) in db.keys("push-device:")? {
        let subscription = key.trim_start_matches("push-device:");
        let key = format!("push-outbox:{}:{subscription}", text(question, "id"));
        if db.kv(&key)?.is_none() {
            db.set(
                &key,
                &json!({
                "subscriptionId":subscription,"questionId":question["id"],"chatId":question["chatId"],"attempts":0,"nextAt":now(),"expiresAt":now()+3600000}
                ),
                None,
            )?;
        }
    }
    Ok(())
}
fn subscription(input: Value) -> Result<Value> {
    let endpoint = text(&input, "endpoint");
    let url =
        url::Url::parse(endpoint).map_err(|_| Error::bad("Invalid notification subscription."))?;
    let host = url.host_str().unwrap_or("");
    let allowed = [
        "fcm.googleapis.com",
        "updates.push.services.mozilla.com",
        "web.push.apple.com",
    ]
    .contains(&host)
        || [".push.services.mozilla.com", ".notify.windows.com"]
            .iter()
            .any(|suffix| host.ends_with(suffix));
    if endpoint.len() > 4096
        || url.scheme() != "https"
        || !url.username().is_empty()
        || url.password().is_some()
        || url.port().is_some()
        || url.fragment().is_some()
        || !allowed
    {
        return Err(Error::bad("This browser push service is not supported."));
    }
    for (key, length, maximum) in [("p256dh", 65, 100), ("auth", 16, 30)] {
        let value = text(&input["keys"], key);
        if value.len() > maximum
            || URL_SAFE_NO_PAD
                .decode(value.trim_end_matches('='))
                .map_or(true, |bytes| bytes.len() != length)
        {
            return Err(Error::bad("Invalid notification subscription keys."));
        }
    }
    Ok(json!({
    "endpoint":endpoint,"keys":{
    "p256dh":input["keys"]["p256dh"],"auth":input["keys"]["auth"]}
    }
    ))
}
impl Notifications {
    async fn keys(&self, service: &Service) -> Result<Value> {
        let vault = service.vault.clone();
        service
            .store
            .transaction(move |db| {
                if let Some(value) = db.kv("mcp-secret:push-vapid")? {
                    return vault.decrypt("push-vapid", &value);
                }
                let private = p256::SecretKey::random(&mut p256::elliptic_curve::rand_core::OsRng);
                let value = json!({
                "privateKey":URL_SAFE_NO_PAD.encode(private.to_bytes()),"publicKey":URL_SAFE_NO_PAD.encode(private.public_key().to_encoded_point(false).as_bytes())}
                );
                vault.set_in(db, "push-vapid", &value)?;
                Ok(value)
            })
            .await
    }
    pub async fn configuration(&self, service: &Service) -> Result<Value> {
        Ok(json!({
        "publicKey":self.keys(service).await?["publicKey"]}
        ))
    }
    pub async fn subscribe(&self, service: &Service, input: Value) -> Result<Value> {
        let value = subscription(input)?;
        let id = hex_digest(text(&value, "endpoint"));
        let vault = service.vault.clone();
        service
            .store
            .transaction(move |db| {
                let key = format!("push-device:{id}");
                if db.kv(&key)?.is_none() && db.keys("push-device:")?.len() >= 50 {
                    return Err(Error::new(
                        409,
                        "Too many notification devices are registered.",
                    ));
                }
                vault.set_in(db, &key, &value)?;
                db.set(
                    &key,
                    &json!({
                    "createdAt":now()}
                    ),
                    None,
                )?;
                Ok(json!({
                "id":id}
                ))
            })
            .await
    }
    pub async fn unsubscribe(&self, service: &Service, id: &str) -> Result<Value> {
        let id = id.to_owned();
        service
            .store
            .transaction(move |db| {
                db.delete(&format!("mcp-secret:push-device:{id}"))?;
                db.delete(&format!("push-device:{id}"))?;
                for (key, value) in db.keys("push-outbox:")? {
                    if value["subscriptionId"] == id {
                        db.delete(&key)?;
                    }
                }
                Ok(json!({
                "ok":true}
                ))
            })
            .await
    }
    pub async fn flush(&self, service: &Service) -> Result<()> {
        let Ok(_delivery) = self.delivery.try_lock() else {
            return Ok(());
        };
        for (key, mut delivery) in service.store.keys("push-outbox:").await? {
            if delivery["nextAt"].as_i64().unwrap_or(0) > now() {
                continue;
            }
            let question = service
                .store
                .kv(&format!(
                    "chat-question:{}:{}",
                    text(&delivery, "chatId"),
                    text(&delivery, "questionId")
                ))
                .await?;
            let subscription = service
                .vault
                .get(&format!(
                    "push-device:{}",
                    text(&delivery, "subscriptionId")
                ))
                .await?;
            if subscription.is_none()
                || question.is_none_or(|question| question["status"] != "pending")
                || delivery["expiresAt"].as_i64().unwrap_or(0) < now()
            {
                service.store.delete(&key).await?;
                continue;
            }
            let subscription = subscription.unwrap();
            // Recheck the allowlist before any network request, including migrated records.
            if self::subscription(subscription.clone()).is_err() {
                self.unsubscribe(service, text(&delivery, "subscriptionId"))
                    .await?;
                continue;
            }
            let keys = self.keys(service).await?;
            let result = tokio::time::timeout(
                Duration::from_secs(5),
                send(service, &subscription, &keys, &delivery),
            )
            .await;
            match result {
                Ok(Ok(())) => service.store.delete(&key).await?,
                Ok(Err(WebPushError::EndpointNotValid(_) | WebPushError::EndpointNotFound(_))) => {
                    self.unsubscribe(service, text(&delivery, "subscriptionId"))
                        .await?;
                }
                _ => {
                    let attempts = delivery["attempts"].as_u64().unwrap_or(0);
                    delivery["attempts"] = attempts.saturating_add(1).into();
                    delivery["nextAt"] =
                        (now() + (10000_i64 * (1_i64 << attempts.min(5))).min(300000)).into();
                    service.store.set(&key, delivery, None).await?;
                }
            }
        }
        Ok(())
    }
}
async fn send(
    service: &Service,
    subscription: &Value,
    keys: &Value,
    delivery: &Value,
) -> std::result::Result<(), WebPushError> {
    let info = SubscriptionInfo::new(
        text(subscription, "endpoint"),
        text(&subscription["keys"], "p256dh"),
        text(&subscription["keys"], "auth"),
    );
    let mut signature = VapidSignatureBuilder::from_base64(text(keys, "privateKey"), &info)?;
    signature.add_claim(
        "sub",
        if service.config.public_url.starts_with("https:") {
            &service.config.public_url
        } else {
            "mailto:notifications@example.com"
        },
    );
    let payload = json!({
    "title":"Your agent has a question","body":"Open the chat to answer.","chatId":delivery["chatId"],"questionId":delivery["questionId"]}
    )
    .to_string();
    let mut message = WebPushMessageBuilder::new(&info);
    message.set_vapid_signature(signature.build()?);
    message.set_ttl(3600);
    message.set_urgency(Urgency::High);
    message.set_payload(ContentEncoding::Aes128Gcm, payload.as_bytes());
    HyperWebPushClient::new().send(message.build()?).await
}
